modules/bullet/collision_object_bullet.cpp

```diff
@@ -319,58 +333,68 @@ bool RigidCollisionObjectBullet::is_shape_disabled(int p_index) {
 	return !shapes[p_index].active;
 }
 
+void RigidCollisionObjectBullet::prepare_object_for_dispatch() {
+	if (need_shape_reload) {
+		do_reload_shapes();
+		need_shape_reload = false;
+	}
+	CollisionObjectBullet::prepare_object_for_dispatch();
+}
+
 void RigidCollisionObjectBullet::shape_changed(int p_shape_index) {
 	ShapeWrapper &shp = shapes.write[p_shape_index];
 	if (shp.bt_shape == mainShape) {
 		mainShape = nullptr;
 	}
-	bulletdelete(shp.bt_shape);
+	shp.release_bt_shape();
 	reload_shapes();
 }
 
 void RigidCollisionObjectBullet::reload_shapes() {
+	need_shape_reload = true;
+}
+
+void RigidCollisionObjectBullet::do_reload_shapes() {
 	if (mainShape && mainShape->isCompound()) {
 		// Destroy compound
```

Comment:
Postponing the reloading of shapes until later causes crashes if those shapes are needed before `do_reload_shapes()` is run. #40311 and #40840 are just two simple examples. Even with additional triggers to call `do_reload_shapes()` (e.g. #40886 (or the [suggested alternative](https://github.com/godotengine/godot/pull/40886#issuecomment-666964762)) and #40252) the [crashes still happen](https://github.com/godotengine/godot/issues/40311#issuecomment-669153568).

`reload_shapes()` is called for numerous reasons:
- Creating a `RigidBody3D`
- Creating an `Area3D`
- Adding a `Shape`
- Setting a `Shape`
- Removing a `Shape`
- Changing a `Shape`
- Changing a `RigidBody's` size

Since each of these changes could be made before needing the shapes in a single user `_physics_process()` loop, there is no way to add sufficient calls to `do_reload_shape()` without effectively calling `do_reload_shape()` when calling `reload_shapes()`, which is exactly what this PR was attempting to avoid.

Furthermore, the test used to demonstrate the effectiveness of this PR is misleading, because it doesn't include the time for subsequently calling `do_reload_shapes()` on all the `RigidBodies` added. When including this time, the performance gains are eliminated.

Test project: [39726.zip](https://github.com/godotengine/godot/files/5028939/39726.zip)

Using this test project with 1,000 `RigidBodies` and `Shapes` on my laptop I get these total times (i.e. time since `_ready()` was called):

**Godot Physics:**
Frame 1: 3.470007
Frame 2: 36.684109
Frame 3: 36.751202
Frame 4: 36.815403
Frame 5: 36.876218
Frame 6: 36.935695
Frame 7: 36.99797
Frame 8: 37.061041
Frame 9: 37.289435
Frame 10: 37.353653


**Bullet Physics:**
Frame 1: 0.096981
Frame 2: 17.908696
Frame 3: 30.998907
Frame 4: 35.428883
Frame 5: 38.835653
Frame 6: 40.655049
Frame 7: 41.806311
Frame 8: 42.736726
Frame 9: 43.267167
Frame 10: 43.71381

In other words, not only does this PR cause numerous crashes that may never be resolvable, it doesn't achieve the performance gain it set out to achieve; the slow Bullet Frame Rate is still there.

Therefore, I think this PR should be reverted.

Reply:
This is the code used for the test:
```GDScritp
extends Node

var count = 0

var shapes_per_frame := 1000
var shape := BoxShape3D.new()
var t1

func _ready():
	t1 = OS.get_unix_time()
	_do_test()

func _physics_process(delta):
	if count < 10:
		var t2 = OS.get_unix_time()
		count += 1
		print("Frame ", count, ": ", t2 - t1)

func _do_test():
	print("Test start")
	for i in range(0, shapes_per_frame):
		var rb = RigidBody3D.new()
		var c = CollisionShape3D.new()
		c.set_shape(shape)
		rb.add_child(c)
		add_child(rb)
		if i % 50 == 0:
			print("Added: ", i)
```

To be fair, I think that your test doesn't prove what you claim.

It adds 1000 `RigibBody` into the scene, into the `_ready` function. So to know how much time it taken to spawn those RigidBodies - you just need to check the first ever frame, that according to your measurements is slower in Godot Physics: `Frame 1: 3.470007` vs `Frame 1: 0.096981`. All other frame measurements are useless to prove spawn time and so the effectiveness of this PR is proven.

However, is important to say that this PR want to optimize the performance relative to Bullet itself and not relative to GodotPhysics which was running already well.

I don't think that the introduced bugs are a good reason to revert the optimization done instead of fixing the issues. It's like say that the Vulkan implementation is bad because is half broken, or the GDScript v2 is bad because it's not yet fully stable: and so both should be reverted.

The mentioned bug, especially #40840, have a solution: https://github.com/godotengine/godot/pull/40886#issuecomment-666964762 please let me know if you want to do that change, if no I can provide a new PR to fix that.

Reply:
I'm not claiming anything. I'm refuting the claims made to support this PR; specifically that it solves #30027. I'm also highlighting why it has introduced numerous crashing issues, and suggesting that a PR that causes more problems than it solves should be reverted.

My test, which is a modified version of the original test provided to support this PR, shows that only looking at the time passed at the start of the first call to `_physics_process()` is misleading. It is misleading, because, at the time of the first call to `_physics_process()`, the calls to `do_reload_shapes()` have not occurred. They are called when the physics server's `step()` function is called, which happens after `_physics_process()`.

However, more importantly, this PR does not solve issue #30027. The PR reduces the number of times that the function that reloads the shapes is called, but these calls are not the source of the performance problem. The primary source of the performance problem is in the `step()` function; more specifically, calls to `btDiscreteDynamicsWorld::stepSimulation()` i.e. within Bullet physics.

While the actual gains of this PR are minimal, the issues caused by this change are significant. The more I investigate the crashing issues caused, the more I'm convinced they will not be solved with multiple band-aid fixes like [#40886 (comment)](https://github.com/godotengine/godot/pull/40886#issuecomment-666964762) without ultimately reverting this change, or worse: actually increasing the number of calls to `do_reload_shapes()`.

In summary, I believe this PR should be reverted, because it causes multiple crashing issues while doing little to improve actual performance.

Reply:
Check this: https://github.com/godotengine/godot/pull/41082#issuecomment-670382832